src/node_buffer.cc

```diff
@@ -583,10 +584,20 @@ void DecodeUTF8(const FunctionCallbackInfo<Value>& args) {
   ArrayBufferViewContents<char> buffer(args[0]);
 
   bool ignore_bom = args[1]->IsTrue();
+  bool has_fatal = args[2]->IsTrue();
 
   const char* data = buffer.data();
   size_t length = buffer.length();
 
+  if (has_fatal) {
+    auto result = simdutf::validate_utf8_with_errors(data, length);
```

Comment:
I've read https://github.com/nodejs/node/pull/45803#issuecomment-1358337268 but I don't understand. Shouldn't `simdutf::validate_utf8()` be faster?

Reply:
According to `simdutf` documentation:

```
 * Validate the UTF-8 string and stop on error. It might be faster than
 * validate_utf8 when an error is expected to occur early.
```

Reply:
How can you expect an error to occur early if you don't know the input?

Reply:
@lpinca From the simdutf README:

> Generally speaking, functions that report errors always stop soon after an error is encountered and might therefore be faster on inputs where an error occurs early in the input. The functions that return a boolean indicating whether or not an error has been encountered are meant to be used in an optimistic setting---when we expect that inputs will almost always be correct.

So if the purpose is to catch content that is garbage, then `validate_utf8_with_errors` will be typically faster. It is an optimization issue ultimately on the number of branches you use.  If you expect that you will read the whole string anyhow, you just collect a flag and you do not repeatedly branch on it, you wait at the end... If you expect that the input will be garbage most of the time, then you want to be branching as much as possible.

Reply:
We don't but if we are using `fatal`, IMHO we are preparing ourselves for the faulty decoding, and hence the probability of having an error outweighs the probability of truth values. If it was a `isUtf8` function, `simdutf::validate_utf8` would have made sense more, where the probability of false and true values should be treated the same.

If we don't come to a consensus, I'm OK with changing this to a `simdutf::validate_utf8`, but that would once again rise @jasnell 's [comment](https://github.com/nodejs/node/pull/45803#issuecomment-1357987353)

Reply:
Note that both functions should be really fast if the input is valid. Here are the results on my mac laptop:

```
validate_utf8+arm64, input size: 81685, iterations: 2000, dataset: unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
   1.290 ins/byte,    0.293 cycle/byte,   12.252 GB/s (5.1 %),     3.589 GHz,    4.404 ins/cycle
   2.303 ins/char,    0.523 cycle/char,    6.864 Gc/s (5.1 %)     1.78 byte/char
validate_utf8_with_errors+arm64, input size: 81685, iterations: 2000, dataset: unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
   1.337 ins/byte,    0.301 cycle/byte,   11.881 GB/s (2.3 %),     3.575 GHz,    4.444 ins/cycle
   2.387 ins/char,    0.537 cycle/char,    6.657 Gc/s (2.3 %)     1.78 byte/char
```

So you are validating at >10 GB/s irrespective of which function you choose (here I use Arabic which is not trivial to validate).

Reply:
Same benchmark on a recent Intel processor...

```
validate_utf8+icelake, input size: 81685, iterations: 2000, dataset: Arabic-Lipsum.utf8.txt
  17.454 GB/s (28.5 %)    9.779 Gc/s     1.78 byte/char
validate_utf8_with_errors+icelake, input size: 81685, iterations: 2000, dataset: Arabic-Lipsum.utf8.txt
  15.631 GB/s (30.6 %)    8.757 Gc/s     1.78 byte/char
```

In both cases, you validate Arabic at a speed higher than 15 GB/s. If you expect the input to be valid (99.9% of the case), then go with validate_utf8, but if there is a decent chance that there will be an error, then validate_utf8_with_errors is fine.

Reply:
I think the common case is to have valid UTF8 or at least it is for WebSocket frames.

Reply:
@lpinca For websocket frames, I have a different pull request open where I propose adding `isUtf8`.